src/Stack/Clean.hs

```diff
@@ -42,15 +42,19 @@ clean (CleanOpts targets) = do
                         then Map.elems locals -- default to cleaning all local packages
                         else mapMaybe (`Map.lookup` locals) targets
             forM_ lpvs $ \(LocalPackageView{lpvRoot = pkgDir},_) -> do
-                distDir <- distDirFromDir pkgDir
-                removeTreeIfExists distDir
+                let delDir =
+                          if doFullClean
+                              then workDirFromDir pkgDir
+                              else distDirFromDir pkgDir
+                removeTreeIfExists =<< delDir
```

Comment:
There's a good chance that the per-package `.stack-work`s will be eliminated at some point (they only exist to work around a Cabal issue on Windows).  As such, I'd suggest making it a requirement that `stack clean --full` applies to the whole project (so it would be an error to specify packages).